### PR TITLE
oci: wait until pidfile to be written before starting timer

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:
We want a way to judging how long the actual exec process is running in a container, to be totally fair in how we account for the exec. Unfortunately, there is not a great signal other than pid file creation (which still is a bit early). This improves the time the actual container process gets for exec by about 10ms
#### Does this PR introduce a user-facing change?

```release-note
fix a bug where a container exec process received a little less time than the timeout provided
```

### CODE REVIEW

File: utils/fsnotify.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/cri-o/cri-o/internal/log"
+	"github.com/fsnotify/fsnotify"
+)
+
+func Watch(ctx context.Context, path string, opsToWatch []fsnotify.Op) chan error {
```
Comment (saschagrunert) [inline utils/fsnotify.go:12]: Do you think it makes sense to add some unit test cases for this function?

File: utils/fsnotify.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+					if event.Op&op == op {
+						if event.Name == path {
+							done <- struct{}{}
+							return
+						}
+					}
+				}
+			case err := <-watcher.Errors:
+				if err != nil {
+					internalErrorCh <- fmt.Errorf("watch error %v", err)
```
Comment (saschagrunert) [inline utils/fsnotify.go:40]: I would wrap those errors:
```suggestion
					internalErrorCh <- errors.Wrap(err, "watch failed")
```

File: utils/fsnotify.go (hunk 1 above)
Comment (haircommander) [reply utils/fsnotify.go]: it does!

File: utils/notify.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	externalErrorCh := make(chan error)
+
+	dir := filepath.Dir(path)
+	if err := notify.Watch(dir, c, opsToWatch...); err != nil {
+		externalErrorCh <- err
+		return externalErrorCh
+	}
+	go func() {
+		defer notify.Stop(c)
+		for {
```
[the remark this answers is not in the document]
Comment (haircommander) [reply utils/notify.go]: u right

**Comment by haircommander** [comment]
I've opted to move to github.com/rjeczalik/notify  instead of fsnotify, partially because it seems more active, and also because I was having trouble with the unit tests and the accuracy of the cross-platform events. As a result of the new package, I have been able to shave off another ~30 ms, leaving the runc overhead at about 4ms on my machine, as opposed to 50 ms before this PR

**Comment by haircommander** [comment]
I also had the idea to test what the behavior was like when conmon was running execs.

turns out, conmon was very generous to container processes--giving somewhere between 40ms and 50ms over the timeout:
`sudo crictl exec --sync --timeout 1 8d sleep 1.05`

in other words, this PR changes us from giving 100ms less from pre-conmonless-exec days to about 50ms less. BUT we are technically closer to the actual time the container process is execing

**Comment by haircommander** [comment]
dropping a484419d9 for now to see if that broke stuff

File: go.mod [hunk 4]
```diff
@@ -49,6 +49,7 @@ require (
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus/client_golang v1.11.0
 	github.com/psampaz/go-mod-outdated v0.8.0
+	github.com/rjeczalik/notify v0.9.2
```
Comment (saschagrunert) [inline go.mod:52]: Do we wanna switch to this dependency completely instead of `github.com/fsnotify/fsnotify`?

File: internal/oci/runtime_oci.go [hunk 5]
```diff
@@ -1099,3 +1074,24 @@ func prepareProcessExec(c *Container, cmd []string, tty bool) (processFile strin
 func (c *Container) conmonPidFilePath() string {
 	return filepath.Join(c.bundlePath, "conmon-pidfile")
 }
+
+func WatchForFile(ctx context.Context, path string, opsToWatch []notify.Event) chan error {
```
Comment (saschagrunert) [inline internal/oci/runtime_oci.go:1078]: this function can be private

Comment (saschagrunert) [inline internal/oci/runtime_oci.go:1078]:
```suggestion
func WatchForFile(ctx context.Context, path string, opsToWatch ...notify.Event) chan error {
```

**Review by saschagrunert — COMMENTED** [review]
Generally LGTM, just a few nits

File: go.mod (hunk 4 above)
Comment (haircommander) [reply go.mod:52]: I found there were test failures with it and I want to jam this through first.

File: internal/oci/runtime_oci.go (hunk 5 above)
Comment (haircommander) [reply internal/oci/runtime_oci.go:1078]: not if we want to unit test it, right?

File: internal/oci/runtime_oci.go [hunk 6]
```diff
@@ -1099,3 +1055,26 @@ func prepareProcessExec(c *Container, cmd []string, tty bool) (processFile strin
...
+
+	dir := filepath.Dir(path)
+	if err := notify.Watch(dir, c, opsToWatch...); err != nil {
+		externalErrorCh <- err
+		return externalErrorCh
+	}
+	go func() {
+		defer notify.Stop(c)
+		for ei := range c {
+			fmt.Println(ei)
```
Comment (haircommander) [inline internal/oci/runtime_oci.go]: todo gotta drop this

File: internal/oci/runtime_oci.go (hunk 5 above)
Comment (haircommander) [reply internal/oci/runtime_oci.go:1078]: todo ctx is unused

Comment (saschagrunert) [reply internal/oci/runtime_oci.go]: Uh, yeah right.

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Comment by haircommander** [comment]
```
time="2021-07-16T10:18:11Z" level=error msg="Error copying docker://quay.io/crio/oom:latest to dir:/go/src/github.com/cri-o/cri-o/.artifacts/oom-image: Error reading blob sha256:131fcea386e736dfaf075424d7e2a31ae8a7fcbb9a87ba93c8a70edd120b
Error pulling docker://quay.io/crio/oom:latest
```

quay flake
/retest

**Comment by haircommander** [comment]
I dropped the bats test, because it was flaky, and I added some code to make sure we're not leaking channels or routines.

**Pull request merged**